applications/StructuralMechanicsApplication/custom_utilities/multipoint_constraint_data.hpp

```diff
@@ -0,0 +1,246 @@
+permit persons to whom the Software  is furnished to do so, subject to
+the following condition:
+
+Distribution of this code for  any  commercial purpose  is permissible
+ONLY BY DIRECT ARRANGEMENT WITH THE COPYRIGHT OWNERS.
+
+The  above  copyright  notice  and  this permission  notice  shall  be
+included in all copies or substantial portions of the Software.
+
+THE  SOFTWARE IS  PROVIDED  "AS  IS", WITHOUT  WARRANTY  OF ANY  KIND,
+EXPRESS OR  IMPLIED, INCLUDING  BUT NOT LIMITED  TO THE  WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+IN NO EVENT  SHALL THE AUTHORS OR COPYRIGHT HOLDERS  BE LIABLE FOR ANY
+CLAIM, DAMAGES OR  OTHER LIABILITY, WHETHER IN AN  ACTION OF CONTRACT,
+TORT  OR OTHERWISE, ARISING  FROM, OUT  OF OR  IN CONNECTION  WITH THE
+SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+==============================================================================
+*/
+//
+//   Project Name:        Kratos
+//   Last Modified by:    $Author: Aditya Ghantasala $
+//   Date:                $Date: 14-03-2017 $
+//   Revision:            $Revision: 1.00 $
+//
+//
+
+#if !defined(CONSTRAINT_SLAVE_H_INCLUDED)
+#define CONSTRAINT_SLAVE_H_INCLUDED
+
+#include <vector>
+#include <map>
+#include <iostream>
+
+namespace Kratos
+{
+
+/** \brief Quaternion
+	* A simple class that implements the main features of quaternion algebra
+	*/
+class MpcData : public Flags
+{
+
+  public:
+    typedef Dof<double>* DofPointerType;
+    typedef Dof<double> DofType;
+
+    ///@name Life Cycle
+    ///@{
+
+    /**
+		Creates a Quaternion from its coefficients.
+		@param id : The ID of the DOF which is considered as a slave
+		*/
+    MpcData()
+    {
+        this->totalNumberOfMasterDOFs = 0;
+    }
+    /// Destructor.
+    virtual ~MpcData(){};
+
+    ///@}
+
+  public:
+    ///@name Operators
+    ///@{
+
+    ///@}
+
+  public:
+    ///@name Access
+    ///@{
+
+    /**
+		Adds a master and a corresponding weight to the current slave
+		@param slaveDOF the DOF number which is to be added
+		*/
+    void AddSlave(DofPointerType slaveDOF)
+    {
+
+        if (this->slaveDofMasterDataMap.count(slaveDOF) > 0)
+        {
+            std::cout << "AddSlave :: Slave DOF already in the map ... skipping." << std::endl;
+        }
+        else
+        {
+            slaveDofVector.push_back(slaveDOF);
+            masterData p;
+            this->slaveDofMasterDataMap[slaveDOF] = p;
+        }
+    }
+    /**
+		Adds a slave to the data structure
+		@param slaveDOF the slave DOF to which master DOF is to be added
+		@param masterDOF the DOF of the master DOF which is to be added
+		@param masterWeight the weight of the masterDOF
+		*/
+    void AddMaster(DofPointerType slaveDOF, DofPointerType masterDOF, double masterWeight)
+    {
+        this->slaveDofMasterDataMap[slaveDOF].masterDOFs.push_back(masterDOF);
+        this->slaveDofMasterDataMap[slaveDOF].masterWeights.push_back(masterWeight);

```

Comment:
it would be slightly better for performance if you do

     auto & sd = this->slaveDofMasterDataMap[slaveDOF];
     sd.masterDOFs.push_back(masterDOF);
     sd.masterWeights.push_back(masterWeight);

on a different note, i am a bit perplex that you use push_back here. what if the master dof is already there?